packages/syft/src/syft/service/code/user_code.py

```diff
@@ -862,24 +977,195 @@ class UserCodeExecutionResult(SyftObject):
+
+        self.job_set_n_iters = job_set_n_iters
+        self.job_set_current_iter = job_set_current_iter
+        self.job_increase_current_iter = job_increase_current_iter
+        self.launch_job = launch_job
+        self.is_async = context.job is not None
+
+
+def execute_byte_code(
+    code_item: UserCode, kwargs: Dict[str, Any], context: AuthedServiceContext
+) -> Any:
     stdout_ = sys.stdout
     stderr_ = sys.stderr
 
     try:
+        # stdlib
+        import builtins as __builtin__
+
+        original_print = __builtin__.print
+
+        safe_context = SecureContext(context=context)
+
+        class LocalDomainClient:
+            def init_progress(self, n_iters):
+                if safe_context.is_async:
+                    safe_context.job_set_current_iter(0)
+                    safe_context.job_set_n_iters(n_iters)
+
+            def set_progress(self, to) -> None:
+                self._set_progress(to)
+
+            def increment_progress(self, n=1) -> None:
+                self._set_progress(by=n)
+
+            def _set_progress(self, to=None, by=None):
+                if safe_context.is_async is not None:
+                    if by is None and to is None:
+                        by = 1
+                    if to is None:
+                        safe_context.job_increase_current_iter(current_iter=by)
+                    else:
+                        safe_context.job_set_current_iter(to)
+
+            @final
+            def launch_job(self, func: UserCode, **kwargs):
+                return safe_context.launch_job(func, **kwargs)
+
+            def __setattr__(self, __name: str, __value: Any) -> None:
+                raise Exception("Attempting to alter read-only value")
+
+        if context.job is not None:
+            job_id = context.job_id
+            log_id = context.job.log_id
+
+            def print(*args, sep=" ", end="\n"):
+                def to_str(arg: Any) -> str:
+                    if isinstance(arg, bytes):
+                        return arg.decode("utf-8")
+                    if isinstance(arg, Job):
+                        return f"JOB: {arg.id}"
+                    if isinstance(arg, SyftError):
+                        return f"JOB: {arg.message}"
+                    if isinstance(arg, ActionObject):
+                        return str(arg.syft_action_data)
+                    return str(arg)
+
+                new_args = [to_str(arg) for arg in args]
+                new_str = sep.join(new_args) + end
+                log_service = context.node.get_service("LogService")
+                log_service.append(context=context, uid=log_id, new_str=new_str)
+                time = datetime.datetime.now().strftime("%d/%m/%y %H:%M:%S")
+                return __builtin__.print(
+                    f"{time} FUNCTION LOG ({job_id}):",
+                    *new_args,
+                    end=end,
+                    sep=sep,
+                    file=sys.stderr,
+                )
+
```

Comment:
A few key points:
- This is only capturing logs generated from print function & not logging std lib.
- For each print() we are triggering a Mongo API call - any failure exception in this API or downstream monog server outage results in direct failure of the function and the job even though the function itself has no issues.
- There's no rate limit on the add/apend API call itself hence a `while(True): print(...)` can probably overwhelm the single-node mongo server.

I would recommend that we setup logging & patch print to write to a file in /var/log instead. Once the function completes (or errors out) we simply upload the logs in a single Mongo API call.

Reply:
yes those are definitely next steps. Not sure about writing the logs only after finishing. The whole point of logs is that you can see them as the program runs. This is especially useful if its a long running job